flang/lib/Lower/CharacterRuntime.cpp

```diff
@@ -157,15 +157,19 @@ void Fortran::lower::genTrim(Fortran::lower::FirOpBuilder &builder,
   builder.create<fir::CallOp>(loc, trimFunc, args);
 }
 
-/// generate call to Scan runtime library routine. 
-/// TODO: Add specialized kind versions.
+/// Generate call to scan or verify runtime routine.
+/// This calls the descriptor based runtime call implementation of the scan
+/// or verify intrinsic.
 void
-Fortran::lower::genScan(Fortran::lower::FirOpBuilder &builder,
-                             mlir::Location loc, mlir::Value resultBox,
-                             mlir::Value stringBox, mlir::Value setBox,
-                             mlir::Value backBox, mlir::Value kind) {
-  auto scanFunc = getRuntimeFunc<mkRTKey(Scan)>(loc, builder);
-  auto fTy = scanFunc.getType();
+Fortran::lower::genScanVerify(Fortran::lower::FirOpBuilder &builder,
+                              mlir::Location loc, mlir::Value resultBox,
+                              mlir::Value stringBox, mlir::Value setBox,
+                              mlir::Value backBox, mlir::Value kind, 
+                              bool isScan) {
+
+  auto func = isScan ? getRuntimeFunc<mkRTKey(Scan)>(loc, builder)
+                     : getRuntimeFunc<mkRTKey(Verify)>(loc, builder);
```

Comment:
The different entry points can all share a common implementation that takes the specific getRuntimeFunc as an argument.